polkadot/node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -2110,6 +2155,223 @@ async fn handle_collation_fetch_response(
+	// parent:
+	// - at RP1 we have got two claims and para A is in the first position in the claim queue. ACC =
+	//   2 - 1 = 1.
+	// - at RP2 we have got no claims advertised. Para A is not on the first position of the claim
+	//   queue, so A2 can't be scheduled here. ACC is still 1.
+	// - at RP 3 we reach our target relay parent with ACC = 1. We want to advertise at RP 3 so we
+	//   add 1 to ACC and it becomes 2. IMPORTANT: since RP3 is our target relay parent we DON'T
+	//   subtract 1 since this is the position the new advertisement will occupy.
+	// - ACC = 2 is the final result which represents the number of claims for para A at RP3.
+	//
+	// It's outside the scope of this function but in this case we CAN add an advertisement for para
+	// A at RP3 since the number of claims equals the number of entries for A in the claim queue.
+
+	// We take all ancestors returned from `known_allowed_relay_parents_under` (remember they start
+	// with `relay_parent`), take the number of elements we are interested in (claim_queue_len) and
+	// then reverse them so that we start from the oldest one.
+	// At each ancestor we take the number of seconded and pending candidates at it, add it to the
+	// total number of claims and subtract 1 if `para_id` is at the first position of the claim
+	// queue for that ancestor.
+	ancestors.iter().take(claim_queue_len).rev().fold(0, |acc, ancestors| {
+		let seconded_and_pending = state.seconded_and_pending_for_para(ancestors, para_id);
+		let first_assignment = state
+			.per_relay_parent
+			.get(ancestors)
+			.and_then(|per_relay_parent| per_relay_parent.assignment.current.first());
+		match first_assignment {
+			Some(first) if first == para_id && ancestors != relay_parent =>
+				(acc + seconded_and_pending).saturating_sub(1),
+			_ => acc + seconded_and_pending,
+		}
+	})
+}
+
+fn seconded_and_pending_for_para_above(
+	state: &State,
+	relay_parent: &Hash,
+	para_id: &ParaId,
+	claim_queue_len: usize,
+) -> Vec<usize> {
+	let leaf_paths = state.implicit_view.paths_to_relay_parent(relay_parent);
+	let mut result = vec![];
+	for path in leaf_paths {
+		let r = path
+			.iter()
+			.take(claim_queue_len)
+			.map(|anc| state.seconded_and_pending_for_para(anc, para_id))
+			.sum();
+		result.push(r);
+	}
+
+	result
+}
+// Returns the claim queue without fetched or pending advertisement. The resulting `Vec` keeps the
+// order in the claim queue so the earlier an element is located in the `Vec` the higher its
+// priority is.
+fn unfulfilled_claim_queue_entries(relay_parent: &Hash, state: &State) -> Result<Vec<ParaId>> {
+	let relay_parent_state = state
+		.per_relay_parent
+		.get(relay_parent)
+		.ok_or(Error::RelayParentStateNotFound)?;
+	let scheduled_paras = relay_parent_state.assignment.current.iter().collect::<HashSet<_>>();
+	let mut claims_per_para = HashMap::new();
+	for para_id in scheduled_paras {
+		let below = seconded_and_pending_for_para_below(
```

Comment:
as discussed offline, we should be using the same `claimed_within_view` function here

Reply:
Fixed